File: snapsmobile/src/main/java/com/snaps/mobile/service/ai/ExifUtil.java
package com.snaps.mobile.service.ai;

import android.media.ExifInterface;

import com.snaps.common.utils.log.Dlog;

import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

/**
 * 파일에서 Exif 정보를 구한다.
 */
class ExifUtil {
    private static final String TAG = ExifUtil.class.getSimpleName();
    //GPS 시간은 왜 단일 포맷이 없는지...
    private static final List<SimpleDateFormat> GPS_SimpleDateFormatList = Arrays.asList(
            new SimpleDateFormat("yyyy:MM:dd HH:mm:ss"),    //이게 경우의 수가 제일 많다.
            new SimpleDateFormat("yyyy-MM-dd HH:mm:ss"),
            new SimpleDateFormat("dd/MM/yyyy HH:mm"),
            new SimpleDateFormat("yyyy:MM:dd:HH:mm:ss"),
            new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
            );

    private long mMinValidDateLong;
    private long mMaxValidDateLong;

    public ExifUtil() {
    }

    public void init() {
        initValidDate();
    }

    /**
     * Exif 정보를 구한다.
     * @param photoInfo
     */
    public void setInfoOnPhotoInfo(PhotoInfo photoInfo) {
        //일단 기본값 설정
        photoInfo.setExifOrientation(PhotoInfo.NOT_EXIST_ORIENTATION_VALUE);
        photoInfo.setExifLatitude(PhotoInfo.NOT_EXIST_GPS_VALUE);
        photoInfo.setExifLongitude(PhotoInfo.NOT_EXIST_GPS_VALUE);
        photoInfo.setExifDateTimeLong(PhotoInfo.NOT_EXIST_DATE_TIME_VALUE);

        try {
            ExifInterface exifInterface = new ExifInterface(photoInfo.getFilePath());
            int orientation = getOrientation(exifInterface);
            float latitude = getLatitude(exifInterface);
            float longitude = getLongitude(exifInterface);
            long dateTimeLong = getDateTimeLong(exifInterface);

            photoInfo.setExifOrientation(orientation);
            photoInfo.setExifLatitude(latitude);
            photoInfo.setExifLongitude(longitude);
            photoInfo.setExifDateTimeLong(dateTimeLong);

        }catch(Exception e) {
            Loggg.e(TAG, e);
        }
    }

    /**
     * 위도 값을 구한다.
     * @param exifInterface
     * @return
     */
    private float getLatitude(ExifInterface exifInterface) {
        try {
            String attrLATITUDE = exifInterface.getAttribute(ExifInterface.TAG_GPS_LATITUDE);
            if (attrLATITUDE == null || attrLATITUDE.length() == 0) return PhotoInfo.NOT_EXIST_GPS_VALUE;

            String attrLATITUDE_REF = exifInterface.getAttribute(ExifInterface.TAG_GPS_LATITUDE_REF);
            if (attrLATITUDE_REF == null || attrLATITUDE_REF.length() == 0) return PhotoInfo.NOT_EXIST_GPS_VALUE;

            float latitude = convertRationalLatLonToFloat(attrLATITUDE, attrLATITUDE_REF);
            return latitude;
        }catch(Exception e) {
            Loggg.e(TAG, e);
            return PhotoInfo.NOT_EXIST_GPS_VALUE;
        }
    }

    /**
     * 경도 값을 구한다.
     * @param exifInterface
     * @return
     */
    private float getLongitude(ExifInterface exifInterface) {
        try {
            String attrLONGITUDE = exifInterface.getAttribute(ExifInterface.TAG_GPS_LONGITUDE);
            if (attrLONGITUDE == null || attrLONGITUDE.length() == 0) return PhotoInfo.NOT_EXIST_GPS_VALUE;

            String attrLONGITUDE_REF = exifInterface.getAttribute(ExifInterface.TAG_GPS_LONGITUDE_REF);
            if (attrLONGITUDE_REF == null ||attrLONGITUDE_REF.length() == 0) return PhotoInfo.NOT_EXIST_GPS_VALUE;

            float longitude = convertRationalLatLonToFloat(attrLONGITUDE, attrLONGITUDE_REF);
            return longitude;
        }catch(Exception e) {
            Loggg.e(TAG, e);
            return PhotoInfo.NOT_EXIST_GPS_VALUE;
        }
    }

    /**
     * 도분초 단위를 도단위로 변경한다.
     * @param rationalString
     * @param ref
     * @return
     */
    private float convertRationalLatLonToFloat(String rationalString, String ref) {
        //https://android.googlesource.com/platform/frameworks/base/+/android-cts-4.4_r1/media/java/android/media/ExifInterface.java
        try {
            String[] parts = rationalString.split(",");
            String[] pair;

            pair = parts[0].split("/");
            double degrees = Double.parseDouble(pair[0].trim()) / Double.parseDouble(pair[1].trim());

            pair = parts[1].split("/");
            double minutes = Double.parseDouble(pair[0].trim()) / Double.parseDouble(pair[1].trim());

            pair = parts[2].split("/");
            double seconds = Double.parseDouble(pair[0].trim()) / Double.parseDouble(pair[1].trim());

            double result = degrees + (minutes / 60.0) + (seconds / 3600.0);
            if ((ref.equals("S") || ref.equals("W"))) {
                return (float) -result;
            }
            return (float)result;
        } catch (Exception e) {
            Loggg.e(TAG, e);
            return PhotoInfo.NOT_EXIST_GPS_VALUE;
        }
    }

    /**
     * 회전 값을 구한다.
     * @param exifInterface
     * @return
     */
    private int getOrientation(ExifInterface exifInterface) {
        try {
            int orientation = exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL);
            return orientation;
        } catch (Exception e) {
            Loggg.e(TAG, e);
            return PhotoInfo.NOT_EXIST_ORIENTATION_VALUE;
        }
    }

    /**
     * 시간 정보를 구한다.
     * @param exifInterface
     * @return
     */
    private long getDateTimeLong(ExifInterface exifInterface) {
        try {
            String dateTimeString = exifInterface.getAttribute(ExifInterface.TAG_DATETIME);
            if (dateTimeString == null || dateTimeString.length() == 0) return PhotoInfo.NOT_EXIST_DATE_TIME_VALUE;

            Date date = null;
            for(SimpleDateFormat sf : GPS_SimpleDateFormatList) {
                try {
                    date = sf.parse(dateTimeString);
                    break;
                }catch (Exception e) {
                    Dlog.e(TAG, e);
                }
            }
            if (date == null) {
                return PhotoInfo.NOT_EXIST_DATE_TIME_VALUE;
            }
            long dateLong = date.getTime();
            dateLong = fixInvalidDate(dateLong);
            return dateLong;
        } catch (Exception e) {
            Loggg.e(TAG, e);
            return PhotoInfo.NOT_EXIST_DATE_TIME_VALUE;
        }
    }


    /**
     * 정상시간 값의 범위를 초기화 한다.
     */
    private void initValidDate() {
        Calendar calendar = Calendar.getInstance();
        calendar.set(1970, 1, 1,0,0,0);
        mMinValidDateLong = calendar.getTimeInMillis(); //그냥 0대입 하면 되는데 혹시 변경에 대비

        calendar = Calendar.getInstance();
        calendar.add(Calendar.MINUTE, 60);
        mMaxValidDateLong = calendar.getTimeInMillis();
    }

    /**
     * 시간이 이상한 경우 한번 수정해보고 그래도 이상하면 0을 리턴
     * @param time
     * @return
     */
    private long fixInvalidDate(long time) {
        //원본 코드 여기 저기 카피 해서 수정
        if (time == 0) return PhotoInfo.NOT_EXIST_DATE_TIME_VALUE;

        if (isValidDate(time)) return time;

        long tmpTime = time * 1000; //sns 사진 중 second로 기록 되어 있는 날짜가 있다.
        if (isValidDate(tmpTime)) return tmpTime;

        return PhotoInfo.NOT_EXIST_DATE_TIME_VALUE;
    }

    /**
     * 시간이 정상인지 검사한다.
     * @param time
     * @return
     */
    private boolean isValidDate(long time) {
        return (mMinValidDateLong <= time && time <= mMaxValidDateLong);
    }
}

File: data/src/main/java/com/snaps/mobile/data/util/ExifUtil.kt
package com.snaps.mobile.data.util

import android.content.ContentResolver
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import com.snaps.common.utils.log.Dlog
import com.snaps.common.android_utils.toOTAngle
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject


class ExifUtil @Inject constructor(
    private val contentResolver: ContentResolver,
) {
    private val tag = ExifUtil::class.java.simpleName

    data class Info(
        val orientation: Int = 0,
        val latitude: Double = 0.0,
        val longitude: Double = 0.0,
        val dateTime: Long = 0
    ) {
        val gps: String?
            get() {
                if (latitude == 0.0 || longitude == 0.0) return null
                return "${latitude},$longitude"
            }

        val exifDate: String
            get() {
                if (dateTime == 0L) return ""
                return SimpleDateFormat("yyyy:MM:dd HH:mm:ss", Locale.KOREA).format(Date(dateTime))
            }

        val orientationAngle: Int
            get() = orientation.toOTAngle()
    }

    //GPS 시간은 왜 단일 포맷이 없는지...
    private val gpsDateFormatList: List<SimpleDateFormat> = listOf(
        SimpleDateFormat("yyyy:MM:dd HH:mm:ss", Locale.KOREA),  //이게 경우의 수가 제일 많다.
        SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.KOREA),
        SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.KOREA),
        SimpleDateFormat("yyyy:MM:dd:HH:mm:ss", Locale.KOREA),
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.KOREA)
    )

    fun getInfo(uriText: String): Info {
        return try {
            contentResolver.openInputStream(Uri.parse(uriText))?.use { inputStream ->
                inputStream?.let {
                    ExifInterface(it)
                }.let {
                    Info(
                        orientation = getOrientation(it),
                        latitude = getLatitude(it),
                        longitude = getLongitude(it),
                        dateTime = getDateTimeLong(it)
                    )
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            Info()
        } ?: Info()
    }

    private fun getDateTimeLong(exifInterface: ExifInterface): Long {
        return exifInterface.getAttribute(ExifInterface.TAG_DATETIME)?.let { dateTimeString ->
            if (dateTimeString.isEmpty()) {
                0L
            } else {
                gpsDateFormatList.firstOrNull {
                    try {
                        it.parse(dateTimeString)
                        true
                    } catch (e: Exception) {
                        false
                    }
                }?.run {
                    parse(dateTimeString).time
                } ?: 0L
            }
        } ?: 0L

//        val dateTimeString = exifInterface.getAttribute(ExifInterface.TAG_DATETIME) ?: return 0
//        if (dateTimeString.isEmpty()) return 0
//
//        for (sf in gpsDateFormatList) {
//            try {
//                return sf.parse(dateTimeString).time
//            } catch (e: Exception) {
//                continue
//            }
//        }
//        return 0
    }

    private fun getOrientation(exifInterface: ExifInterface): Int {
        return try {
            exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL).let {
                when (it) {
                    ExifInterface.ORIENTATION_UNDEFINED -> ExifInterface.ORIENTATION_UNDEFINED
                    ExifInterface.ORIENTATION_NORMAL -> ExifInterface.ORIENTATION_NORMAL
                    ExifInterface.ORIENTATION_ROTATE_90 -> ExifInterface.ORIENTATION_ROTATE_90
                    ExifInterface.ORIENTATION_ROTATE_180 -> ExifInterface.ORIENTATION_ROTATE_180
                    ExifInterface.ORIENTATION_ROTATE_270 -> ExifInterface.ORIENTATION_ROTATE_270
                    //음...
                    ExifInterface.ORIENTATION_TRANSPOSE -> ExifInterface.ORIENTATION_ROTATE_90
                    ExifInterface.ORIENTATION_FLIP_VERTICAL -> ExifInterface.ORIENTATION_ROTATE_180
                    ExifInterface.ORIENTATION_TRANSVERSE -> ExifInterface.ORIENTATION_ROTATE_270
                    else -> ExifInterface.ORIENTATION_UNDEFINED //32380 대응 (unknown)
                }
            }
        } catch (e: Exception) {
            Dlog.e(tag, e)
            ExifInterface.ORIENTATION_UNDEFINED
        }
    }

    private fun getLatitude(exifInterface: ExifInterface): Double {
        return exifInterface.getAttribute(ExifInterface.TAG_GPS_LATITUDE)?.let { latitudeText ->
            exifInterface.getAttribute(ExifInterface.TAG_GPS_LATITUDE_REF)?.let { latitudeRefText ->
                try {
                    convertRationalLatLonToDouble(
                        latitudeText,
                        latitudeRefText
                    )
                } catch (e: Exception) {
                    0.0
                }
            }
        } ?: 0.0
    }

    private fun getLongitude(exifInterface: ExifInterface): Double {
        return exifInterface.getAttribute(ExifInterface.TAG_GPS_LONGITUDE)?.let { longitudeText ->
            exifInterface.getAttribute(ExifInterface.TAG_GPS_LONGITUDE_REF)?.let { longitudeRefText ->
                try {
                    convertRationalLatLonToDouble(
                        longitudeText,
                        longitudeRefText
                    )
                } catch (e: Exception) {
                    0.0
                }
            }
        } ?: 0.0
    }


    //https://android.googlesource.com/platform/frameworks/base/+/android-cts-4.4_r1/media/java/android/media/ExifInterface.java
    //도분초 단위를 도단위로 변경한다.
    private fun convertRationalLatLonToDouble(rationalString: String, ref: String): Double {
        return rationalString.split(",").map {
            it.split("/").let { list ->
                list[0].trim().toDouble() / list[1].trim().toDouble()
            }
        }.toList().let {
            it[0] + it[1] / 60.0 + it[2] / 3600.0 * if (ref == "S" || ref == "W") -1 else 1
        }

//        val parts = rationalString.split(",")
//
//        val degrees = parts[0].split("/").let {
//            it[0].trim().toDouble() / it[1].trim().toDouble()
//        }
//
//        val minutes = parts[1].split("/").let {
//            it[0].trim().toDouble() / it[1].trim().toDouble()
//        }
//
//        val seconds = parts[2].split("/").let {
//            it[0].trim().toDouble() / it[1].trim().toDouble()
//        }
//
//        return degrees + minutes / 60.0 + seconds / 3600.0 * if (ref == "S" || ref == "W") -1 else 1
    }
}